torch/utils/data/dataloader.py

```diff
@@ -190,13 +169,91 @@ def __setattr__(self, attr, val):
         super(DataLoader, self).__setattr__(attr, val)
 
     def __iter__(self):
-        return _DataLoaderIter(self)
+        if self.num_workers == 0:
+            return _SingleProcessDataLoaderIter(self)
+        else:
+            return _MultiProcessingDataLoaderIter(self)
 
     def __len__(self):
-        return len(self.batch_sampler)
+        if self.mode == _DataLoaderStrategy.Iterable:
+            return len(self.dataset)
+        return len(self.sampler)
+
+
+class _BaseDataLoaderIter(object):
```

Comment:
This class has so many `if`s for the iterable dataset, that I'm not sure if it shouldn't be a separate class. If you don't want to repeat all the assignments in `__init__` make them have a superclass, and then override the behavior in subclasses.

Reply:
Yes, I do agree that there is a lot of `if`s here, making it kind of ugly. 

However, I disagree that this should be made into different subclasses. The class inheritance structure is made so that most of the `mode`-dependent code lives in these functions implemented by `_BaseDataLoaderIter`, while `_SingleProcessDataLoaderIter` and `_MultiProcessDataLoaderIter` mostly does mode independent things. Having mode-dependent `DataLoaderIter` classes will be a lot of subclasses, and hurts readability and cleanness IMO.

Reply:
I'm just afraid that right now it is _highly unclear_ which attributes are defined when, and so checking the correctness of the code below if extremely hard. I agree that having 6 subclasses is a bit much, but maybe we can factor out the serial/parallel and iterable/batched/map components and then compose them separately? Right now we lack the second split.

Reply:
I see your point. I'll give it a try to refactor.

Reply:
I did a refactoring where I applied a dummy sampler for `IterableDataset` (which is needed to support batching for its case), and moved the shared logic of actually fetching from dataset to a separate class. Now the `if`-statements on the dataset type is minimal in `_BaseDataLoaderIter` and its subclasses: `1` in total.